File: R/CFU_estimator.R
library(tidyverse)
library(here)

# Global vars -------------------------------------------------------------

# set the poisson threshold N. This should be a value over which you don't trust the counts
# for the 2 ul spotting assay a good value for this is 50
N <- 50

# set the MLE threshold NMLE. A good starting estimate is the ratio of the total area of 
# the plate/spot to the average colony size. Used for calculating MPN. For a 100 mm petri 
# dish a good value is 5000. For a 2 ul spot, a decent value is 100
NMLE <- 100

# the output will be expressed in terms of absolute CFU in the amount that was plated or 
# spotted. It is very important that the same spotting/plating volume be used for
# all dilutions. To convert to CFU/ml divide R by the spot/plating volume. E.g., if these
# are counts from the 96-well spot assay the value is usually 0.002 or 0.0025
sampling_vol <- 0.002

# Functions ---------------------------------------------------------------

find_poisson_cutoff <- function(counts, dil, N, V=1){
  mask <- counts<N
  r_p <- sum(counts[mask])/(sum(dil[mask])*V)
  r_p_std <- sqrt(r_p*r_p/sum(counts[mask]))
  tibble(r = r_p, stderr = r_p_std, type = "poisson_cutoff")
}

find_MLE <- function(counts, dil, N, V=1){
  f <- function(x) {
    sum(counts*dil/(N*(1-exp(-x*dil*V/N))))-sum(dil)
  }
  if (any(N<counts)) {
    tibble(r = Inf, stderr = Inf, type = "ML_estimator")
  } else {
    root <- uniroot(f, interval = c(1, 1e9))
    r_mle <- root$root
    p0 <- exp(-r_mle*dil*V/N)
    invVar <- sum(dil*dil*V*V*counts*p0/(N*N*(1-p0)**2))
    estVar <- 1/invVar
    r_mle_std <- sqrt(estVar)
    tibble(r = r_mle, stderr = r_mle_std, type = "ML_estimator")
  }
}

find_estimators <- function(counts, dil, N, NMLE) {
  bind_rows(find_poisson_cutoff(counts, dil, N),
            find_MLE(counts, dil, NMLE))
}

# Run ---------------------------------------------------------------------

cfu_df <- readr::read_csv(here::here("_data_raw", "CFU_test.csv"), col_names = c("count", "dilution", "group"))

cfu_df %>% 
  nest(data = c(count, dilution)) %>% 
  mutate(r = map(data, \(x) find_estimators(x$count, x$dilution, N=N, NMLE=NMLE))) %>% 
  unnest(r) %>%
  mutate(CFU_ml = r/sampling_vol, 
         CFU_ml_stderr = stderr/sampling_vol) %>% 
  dplyr::select(group, CFU_ml, CFU_ml_stderr, type)

File: Py/CFU_estimator.py
import os
import sys
from pathlib import Path
import pandas as pd
import numpy as np
from scipy import optimize

if (args_count := len(sys.argv)) > 2:
    print(f"One argument expected, got {args_count - 1}")
    raise SystemExit(2)
elif args_count < 2:
    print("You must specify the target CFU csv file")
    raise SystemExit(2)

target_f = Path(sys.argv[1])

if not target_f.is_file():
    print("The target file doesn't exist")
    raise SystemExit(1)

### Find the maximum likelihood estimator for CFUs (MPN method)
### samples - a numpy array of colony counts
### dil - a numpy array of dilutions
### V - the Volume
### N - Max number of Colonies
def findMLE(samples, dil, V, N):
    f = lambda x: np.sum(samples*dil/(N*(1-np.exp(-x*dil*V/N))))-np.sum(dil)
    if any(N<samples):
        return [np.inf, np.inf]
    sol = optimize.root_scalar(f, bracket=[0, 100000000], method='brentq')
    r_mle=sol.root
    p0=np.exp(-r_mle*dil*V/N)
    invVar=np.sum(dil*dil*V*V*samples*p0/(N*N*(1-p0)**2))
    estVar=1/invVar
    r_mle_std=np.sqrt(estVar)
    return [r_mle,r_mle_std]

### Find the Poisson estimator for CFUs
### samples - a numpy array of colony counts
### dil - a numpy array of dilutions
### V - the Volume
def findNaivePoisson(samples, dil, V):
    r_p=np.sum(samples)/(np.sum(dil)*V)
    r_p_std=np.sqrt(r_p*r_p/np.sum(samples))
    return [r_p, r_p_std]

### Find the Poisson estimator with a cutoff
### samples - a numpy array of colony counts
### dil - a numpy array of dilutions
### V - the Volume
### N - Cutoff above which there are crowding effects
def findPoissonCutoff(samples, dil, V, N):
    mask=samples<N
    r_p=np.sum(samples[mask])/(np.sum(dil[mask])*V)
    r_p_std=np.sqrt(r_p*r_p/np.sum(samples[mask]))
    return [r_p, r_p_std]

df = pd.read_csv(target_f, header = None)

print(findMLE(df[0], df[1], 1, 500))
print(findPoissonCutoff(df[0], df[1], 1, 100))